File: src/drivers/python/wappalyzer.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import os
import sys
import PyV8
import requests
from urlparse import urlparse

try:
    import json
except ImportError:
    import simplejson as json


class Wappalyzer(object):

    def __init__(self, url):
        self.file_dir = os.path.dirname(__file__)

        f = open(os.path.join(self.file_dir, 'apps.json'))
        data = json.loads(f.read())
        f.close()

        self.categories = data['categories']
        self.apps = data['apps']
        self.url = url

    def analyze(self):
        ctxt = PyV8.JSContext()
        ctxt.enter()

        f1 = open(os.path.join(self.file_dir, 'js/wappalyzer.js'))
        f2 = open(os.path.join(self.file_dir, '../php/js/driver.js'))
        ctxt.eval(f1.read())
        ctxt.eval(f2.read())
        f1.close()
        f2.close()

        host = urlparse(self.url).hostname
        response = requests.get(self.url)
        html = response.text
        headers = dict(response.headers)

        data = {'host': host, 'url': self.url, 'html': html, 'headers': headers}
        apps = json.dumps(self.apps)
        categories = json.dumps(self.categories)
        return ctxt.eval("w.apps = %s; w.categories = %s; w.driver.data = %s; w.driver.init();" % (apps, categories, json.dumps(data)))

if __name__ == '__main__':
    try:
        w = Wappalyzer(sys.argv[1])
        print w.analyze()
    except IndexError:
        print ('Usage: python %s <url>' % sys.argv[0])

File: src/drivers/ruby/wappalyzer.rb
#!/usr/bin/env ruby

require 'net/http'
require 'v8'
require 'json'

Encoding.default_external = Encoding::UTF_8

class Wappalyzer
  def initialize
    @realdir = File.dirname(File.realpath(__FILE__))
    file = File.join(@realdir, '..', '..', 'share', 'apps.json')
    @json = JSON.parse(IO.read(file))
    @categories, @apps = @json['categories'], @json['apps']
  end

  def analyze(url)
    uri, body, headers = URI(url), nil, {}
    Net::HTTP.start(uri.host, uri.port, :use_ssl => uri.scheme == 'https', :open_timeout => 5) do |http|
      resp = http.get(uri.request_uri)
      resp.each_header{|k,v| headers[k.downcase] = v}
      body = resp.body.encode('UTF-8', :invalid => :replace, :undef => :replace)
    end

    cxt = V8::Context.new
    cxt.load File.join(@realdir, 'js', 'wappalyzer.js')
    cxt.load File.join(@realdir, 'js', 'driver.js')
    data = {'host' => uri.hostname, 'url' => url, 'html' => body, 'headers' => headers}
    output = cxt.eval("w.apps = #{@apps.to_json}; w.categories = #{@categories.to_json}; w.driver.data = #{data.to_json}; w.driver.init();")
    JSON.load(output)
  end
end

if $0 == __FILE__
  url = ARGV[0]
  if url
    puts JSON.pretty_generate(Wappalyzer.new.analyze(ARGV[0]))
  else
    puts "Usage: #{__FILE__} http://example.com"
  end
end
